CLN: named parameters for GroupBy.(mean|median|var|std)

topper-123 opened the pull request:
Drops *args & **kwargs, replace with named parameters for groupby methods mean, median, var & std. Similar to #31473.

This PR has the side effect that the raised error when a parameter is not allowed, is now ``TypeError`` instead of ``UnsupportedFunctionCall``, so technically a API change...

topper-123 committed:
CLN: paramters for GroupBy.(mean|median|var|std)

WillAyd reviewed (commented):
looks good to me. I'd be ok with 1.1 but maybe worth a 2.0 target as well

topper-123 commented:
Mayhbe a decorator similar to ``deprecate_kwargs`` could be added. That way we'd get the correct signature, and ``UnsupportedFunctionCall`` would be raised in the decorator.

jreback commented:
> looks good to me. I'd be ok with 1.1 but maybe worth a 2.0 target as well

we don’t have the bandwidth or infrastructure to handle this type of versioning
if it’s not targeted then would rather close

this is fine for 1.1

topper-123 commented:
Ok, so I don't do a decorator solution?

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1222,14 +1222,15 @@ def mean(self, *args, **kwargs):
     @Substitution(name="groupby")
     @Appender(_common_see_also)
-    def median(self, **kwargs):
+    def median(self, numeric_only=True, min_count=-1):
```
can you update doc-strings for these parameters (or are they already ok with the Appenders?)

jreback reviewed (approved):
lgtm. a question.

topper-123 committed:
Add Parameter doc strings

topper-123 commented:
Ok, I've added doc strings. Also, I've found out that ``Groupby.median`` actually didn't accept ``min_count`` in master, so I've removed the parameter from that method.

jreback commented:
thanks @topper-123

jreback commented:
> Ok, I've added doc strings.
> 
> Also, I've found out that `Groupby.median` actually didn't accept `min_count` in master, but raised an `AssertionError`. So I removed the parameter from that method signature.

happy to have an issue / PR about this.

topper-123 commented:
I did fix that is this PR, so no issue necesary.

So in short, in v1.0.0/master we'd have that ``grp.mean(min_count=1)`` raised a ``AssertionError``, while now it raised a ``TypeError``, which is the same error code as we'd get for ``df.mean(min_count=1)``, so is the correct error type, IMO.
